cpp/include/cudf/join.hpp

```diff
@@ -367,5 +367,138 @@ std::unique_ptr<cudf::table> cross_join(
   cudf::table_view const& right,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Hash join that builds hash table in creation and probes results in subsequent `*_join`
+ * member functions.
+ *
+ * This class enables the hash join scheme that builds hash table once, and probes as many times as
+ * needed (possibly in parallel).
+ */
+class hash_join {
+ public:
+  hash_join() = delete;
+  ~hash_join();
+  hash_join(hash_join const&) = delete;
+  hash_join(hash_join&&)      = delete;
+  hash_join& operator=(hash_join const&) = delete;
+  hash_join& operator=(hash_join&&) = delete;
+
+  /**
+   * @brief Construct a hash join object and build the internal hash table used for subsequent
```

Comment:
When the hash table is built is an implementation detail. E.g., it could be built lazily on the first `*_join` invocation. 

We should also make sure to indicate that this object does not own the `build` table and its lifetime needs to be maintained by the user. 
```suggestion
   * @brief Construct a hash join object for subsequent probe calls. 
   *
   * @note The `hash_join` object must not outlive the table viewed by `build`, else behavior is undefined.
```